sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java

```diff
@@ -656,4 +697,118 @@ public int compareTo(@Nonnull EndPoint o) {
           .result();
     }
   }
+
+  static class BeamRowAdapter implements RowAdapter<com.google.bigtable.v2.Row> {
+    @Override
+    public RowBuilder<com.google.bigtable.v2.Row> createRowBuilder() {
+      return new DefaultRowBuilder();
+    }
+
+    @Override
+    public boolean isScanMarkerRow(com.google.bigtable.v2.Row row) {
+      return Objects.equals(row, com.google.bigtable.v2.Row.getDefaultInstance());
+    }
+
+    @Override
+    public ByteString getKey(com.google.bigtable.v2.Row row) {
+      return row.getKey();
+    }
+
+    private static class DefaultRowBuilder
+        implements RowAdapter.RowBuilder<com.google.bigtable.v2.Row> {
+      private com.google.bigtable.v2.Row.Builder protoBuilder =
+          com.google.bigtable.v2.Row.newBuilder();
+
+      private TreeMap<String, TreeMap<ByteString, ImmutableList.Builder<Cell>>>
+          cellsByFamilyColumn = new TreeMap<>();
+      private TreeMap<ByteString, ImmutableList.Builder<Cell>> cellsByColumn =
+          new TreeMap<>(Comparator.comparing(o -> o.toString(StandardCharsets.UTF_8)));
+      private ImmutableList.Builder<Cell> currentColumnCells;
```

Comment:
I think you are doing a lot of necessary work here....all of the cells will be in contiguous chunks and also you are creating parallel lists that you are throwing away.

Why not just write to proto directly?

```java
startRow() {
  builder = Row.newBuilder().setRow(key);
  lastFamilyName = null;
  lastFamily = null;
  lastColumnName = null;
  lastColumn = null;
}
startCell() {
  bool familyChanged = false;
  if (newFamilyName != lastFamilyName) {
    familyChanged = true;
    lastFamily = builder.newFamilyBuilder().setName(newFamilyName);
    lastFamilyName = newFamilyName;
  }
  if (lastColumnName != newColumnName || familyChanged) {
    lastColumn = lastFamilt.newColumnBuilder().setName(newColumnName);
    lastColumnName = newColumnName;
  }
  lastCell = lastColumn.newCellBuilder().setTimestamp(timestamp).addAllLabels(labels);
  currentValue = null;
}
cellValue() {
  if (currentValue == null) { currentValue = newValue; }
  else currentValue = currentValue.append(newValue);
}
...
```